### PR TITLE
FailureListFormatter: pinpoint failure location and improve editor integration

### PR DESCRIPTION
**Pull request by paddor**
Second attempt of #2839.

### CODE REVIEW

**Comment by paddor** [comment]
This second attempt tries to pinpoint the relevant source of the exception. Aggregated failures produce multiple lines in the output.

**Comment by paddor** [comment]
Fixed CI on Windows.

For more fine-grained matching in editors, I've included an error type (always `E`).

Before:

```
./foobar_spec.rb:12:expected false got true
```

After:
```
./foobar_spec.rb:12:E:expected false got true
```

Without this, it's tricky to safely distinguish these `file:nr:msg` lines from other lines. For example the simple `errorformat='%f:%l:%m'` in Vim would also recognize any line containing a timestamp as an error. Instead, `errorformat='%f:%l:%t:%m'` will only match lines like `path:nr:type:msg`.

This also allows for future extension of this formatter. For example, full backtraces could be printed as informational lines (`I` instead of `E`) if `--backtrace` was given.

**Comment by paddor** [comment]
@JonRowe @smortex

**Comment by paddor** [comment]
So after working with this for a day, I found myself running RSpec without `-ff` for a few times to see more of the backtrace. Thus, I've added support for `--backtrace`.

Users of this `FailureListFormatter` can configure their editors to either run RSpec with `-ff` or with `-ff --backtrace` to get informational (backtrace) lines along with the errors lines where the code failed. It's still the same format so capable editors can jump to the exact locations.

**Comment by paddor** [comment]
I've simplified it more. It always includes backtrace as informational lines. I find this the most useful output, making it very quick to follow how/where an spec failed.

Please review.

File: lib/rspec/core/formatters/failure_list_formatter.rb [hunk 1]
```diff
@@ -17,6 +27,55 @@ def example_failed(failure)
...
+        end
+
+        # @return [String]
+        def one_line(exception)
+          exception.message.lines.map(&:strip).reject(&:empty?).join ' '
+        end
+
+        # @return [Array<String>, nil] relevant location with relative path, if any
+        def find_locations(exception)
+          require 'pathname'
```
Comment (JonRowe) [inline lib/rspec/core/formatters/failure_list_formatter.rb:49]: Sorry but we don't want to include the standard library like this, plus the formatter is showing the example that failed, it doesn't need to be the exception location.

**Review by JonRowe — CHANGES_REQUESTED** [review]

File: .rubocop.yml [hunk 2]
```diff
@@ -46,6 +46,9 @@ Lint/LiteralInInterpolation:
 Lint/NonLocalExitFromIterator:
   Enabled: false
 
+Style/AndOr:
+  EnforcedStyle: conditionals
+
```
[the remark this answers is not in the document]
Comment (paddor) [reply .rubocop.yml:51]: That's what I thought, too. But it's not default. If I run `bundle exec rake rubocop`  without those lines, I get this:

```
Offenses:

lib/rspec/core/formatters/failure_list_formatter.rb:51:43: C: Style/AndOr: Use || instead of or.
          bt        = exception.backtrace or return
                                          ^^
lib/rspec/core/formatters/failure_list_formatter.rb:53:54: C: Style/AndOr: Use || instead of or.
          bt_lines  = bt.reject { |l| exclude =~ l } or return
                                                     ^^
lib/rspec/core/formatters/failure_list_formatter.rb:57:64: C: Style/AndOr: Use || instead of or.
            md            = bt_line.match(/^(.+?):(\d+):(.*)/) or next
                                                               ^^
```

File: lib/rspec/core/formatters/failure_list_formatter.rb (hunk 1 above)
Comment (paddor) [reply lib/rspec/core/formatters/failure_list_formatter.rb]: Honestly, it wasn't my first choice. I wanted to include it the regular way, but another test failed, complaining about the set of set of `require`s that RSpec brings with itself. I moved it back up. Expect to see failed checks.

> plus the formatter is showing the example that failed, it doesn't need to be the exception location.

Could you rephrase that? I don't understand.

Comment (paddor) [reply lib/rspec/core/formatters/failure_list_formatter.rb]: This is the failure that occurs with a regular `require` at the top of the file:

```
Failures:
99

100
  1) RSpec behaves like library wide checks only loads a known set of stdlibs so gem authors are forced to load libs they use to have passing specs
101
     Failure/Error: @example_group_instance.instance_exec(self, &@example_block)
102

103
       expected: []
104
            got: ["/opt/hostedtoolcache/Ruby/2.5.8/x64/lib/ruby/2.5.0/x86_64-linux/pathname.so", "/opt/hostedtoolcache/Ruby/2.5.8/x64/lib/ruby/2.5.0/pathname.rb"]
105

106
       (compared using ==)
107
     Shared Example Group: "library wide checks" called from ./spec/rspec/core_spec.rb:31
108
     # /home/runner/work/rspec-core/rspec-support/lib/rspec/support.rb:97:in `block in <module:Support>'
109
     # /home/runner/work/rspec-core/rspec-support/lib/rspec/support.rb:106:in `notify_failure'
110
     # /home/runner/work/rspec-core/rspec-expectations/lib/rspec/expectations/fail_with.rb:35:in `fail_with'
111
     # /home/runner/work/rspec-core/rspec-expectations/lib/rspec/expectations/handler.rb:38:in `handle_failure'
112
     # /home/runner/work/rspec-core/rspec-expectations/lib/rspec/expectations/handler.rb:56:in `block in handle_matcher'
113
     # /home/runner/work/rspec-core/rspec-expectations/lib/rspec/expectations/handler.rb:27:in `with_matcher'
114
     # /home/runner/work/rspec-core/rspec-expectations/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
115
     # /home/runner/work/rspec-core/rspec-expectations/lib/rspec/expectations/expectation_target.rb:65:in `to'
116
     # /home/runner/work/rspec-core/rspec-support/lib/rspec/support/spec/library_wide_checks.rb:129:in `block (2 levels) in <top (required)>'
117
     # ./lib/rspec/core/example.rb:262:in `instance_exec'
118
     # ./lib/rspec/core/example.rb:262:in `block in run'
119
     # ./lib/rspec/core/example.rb:508:in `block in with_around_and_singleton_context_hooks'
120
     # ./lib/rspec/core/example.rb:465:in `block in with_around_example_hooks'
121
     # ./lib/rspec/core/hooks.rb:486:in `block in run'
122
     # ./lib/rspec/core/hooks.rb:626:in `block in run_around_example_hooks_for'
123
     # ./lib/rspec/core/example.rb:350:in `call'
124
     # ./spec/support/sandboxing.rb:16:in `block (3 levels) in <top (required)>'
125
     # ./lib/rspec/core/sandbox.rb:29:in `sandboxed'
126
     # ./spec/support/sandboxing.rb:7:in `block (2 levels) in <top (required)>'
127
     # ./lib/rspec/core/example.rb:455:in `instance_exec'
128
     # ./lib/rspec/core/example.rb:455:in `instance_exec'
129
     # ./lib/rspec/core/hooks.rb:390:in `execute_with'
130
     # ./lib/rspec/core/hooks.rb:628:in `block (2 levels) in run_around_example_hooks_for'
131
     # ./lib/rspec/core/example.rb:350:in `call'
132
     # ./lib/rspec/core/hooks.rb:629:in `run_around_example_hooks_for'
133
     # ./lib/rspec/core/hooks.rb:486:in `run'
134
     # ./lib/rspec/core/example.rb:465:in `with_around_example_hooks'
135
     # ./lib/rspec/core/example.rb:508:in `with_around_and_singleton_context_hooks'
136
     # ./lib/rspec/core/example.rb:259:in `run'
137
     # ./lib/rspec/core/example_group.rb:644:in `block in run_examples'
138
     # ./lib/rspec/core/example_group.rb:640:in `map'
139
     # ./lib/rspec/core/example_group.rb:640:in `run_examples'
140
     # ./lib/rspec/core/example_group.rb:606:in `run'
141
     # ./lib/rspec/core/example_group.rb:607:in `block in run'
142
     # ./lib/rspec/core/example_group.rb:607:in `map'
143
     # ./lib/rspec/core/example_group.rb:607:in `run'
144
     # ./lib/rspec/core/runner.rb:121:in `block (3 levels) in run_specs'
145
     # ./lib/rspec/core/runner.rb:121:in `map'
146
     # ./lib/rspec/core/runner.rb:121:in `block (2 levels) in run_specs'
147
     # ./lib/rspec/core/configuration.rb:2067:in `with_suite_hooks'
148
     # ./lib/rspec/core/runner.rb:116:in `block in run_specs'
149
     # ./lib/rspec/core/reporter.rb:74:in `report'
150
     # ./lib/rspec/core/runner.rb:115:in `run_specs'
151
     # ./lib/rspec/core/runner.rb:89:in `run'
152
     # ./lib/rspec/core/runner.rb:71:in `run'
153
     # ./lib/rspec/core/runner.rb:45:in `invoke'
154
     # ./exe/rspec:4:in `<top (required)>'
155
     # script/rspec_with_simplecov:42:in `load'
156
     # script/rspec_with_simplecov:42:in `<main>'
```

File: .rubocop.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (JonRowe) [reply .rubocop.yml:51]: Ah I see, no them, please remove this and use `||` we prefer `&&` and `||`

[the remark this answers is not in the document]
Comment (paddor) [reply .rubocop.yml]: Done

File: lib/rspec/core/formatters/failure_list_formatter.rb [hunk 3]
```diff
@@ -1,4 +1,6 @@
 RSpec::Support.require_rspec_core "formatters/base_formatter"
+require 'pathname'
```
Comment (JonRowe) [inline lib/rspec/core/formatters/failure_list_formatter.rb:2]: This can't be merged with any usage of `require 'pathname'` in lib, it's not always available and we don't want false positives for people (where they don't realise they needed to requite it).

Its also unnecessary, no other formatter uses it to look up failure locations. See the exception presenter etc for how we format lines.

File: spec/rspec/core/formatters/failure_list_formatter_spec.rb [hunk 4]
```diff
@@ -4,15 +4,42 @@ module RSpec::Core::Formatters
...
+        |./spec/rspec/core/resources/formatter_specs.rb:50:E:foo
+        |./spec/rspec/core/resources/formatter_specs.rb:50:I:in `...'
+        |./spec/support/formatter_support.rb:41:I:in `...'
+        |./spec/support/formatter_support.rb:3:I:in `...'
+        |./spec/rspec/core/formatters/failure_list_formatter_spec.rb:8:I:in `...'
+        |./spec/support/sandboxing.rb:16:I:in `...'
+        |./spec/support/sandboxing.rb:7:I:in `...'
+        |/foo.html.erb:1:E:Exception
+        |/foo.html.erb:1:I:in `<main>': foo (RuntimeError)
+        |./spec/rspec/core/resources/formatter_specs.rb:71:E:boom
```
Comment (JonRowe) [inline spec/rspec/core/formatters/failure_list_formatter_spec.rb:42]: This massively changes the reason for the formatter, so I'm unsure thats what we want. I'd prefer a much simpler version of this PR that just swaps out the message of the failure from example description to a string from an exception/exceptions (single line per test).

**Review by JonRowe — CHANGES_REQUESTED** [review]

**Comment by pirj** [comment]
Closing as stale
